torchao/csrc/cuda/activation24/sparse_gemm.cu

```diff
@@ -0,0 +1,351 @@
+#include "cutlass/gemm/collective/collective_builder.hpp"
+#include "cutlass/gemm/gemm.h"
+#include "cutlass/gemm/kernel/gemm_universal.hpp"
+#include "cutlass/numeric_types.h"
+#include "cutlass/transform/device/transform_universal_adapter.hpp"
+
+#include <tuple>
+#include <type_traits>
+
+using namespace at;
+
+namespace {
+#define CUTLASS_STATUS_CHECK(status)              \
+  {                                               \
+    TORCH_CHECK(                                  \
+        status == cutlass::Status::kSuccess,      \
+        "Got CUTLASS error: ",                    \
+        cutlass::cutlassGetStatusString(status)); \
+  }
+
+template <typename T>
+struct identity {
+  CUTLASS_HOST_DEVICE
+  T operator()(T lhs) const {
+    return lhs;
+  }
+};
+
+template <typename ElementA>
+struct SparseRowwiseKernel;
+
+template <>
+struct SparseRowwiseKernel<cutlass::float_e4m3_t> {
+  static constexpr auto kElementOutAt = at::ScalarType::BFloat16;
+  static constexpr auto kElementAAt = at::ScalarType::Float8_e4m3fn;
+
+  using ElementA = cutlass::float_e4m3_t;
+  using ElementB = cutlass::float_e4m3_t;
+  using ElementOut = cutlass::bfloat16_t;
+  using ElementAccumulator = float;
+
+  using TileShape = cute::Shape<cute::_128, cute::_256, cute::_128>;
+
+  // Epilogue visitor tree
+  using Accum = cutlass::epilogue::fusion::Sm90AccFetch;
+  using AScale =
+      cutlass::epilogue::fusion::Sm90ColBroadcast<0, TileShape, float>;
+  using BScale =
+      cutlass::epilogue::fusion::Sm90RowBroadcast<0, TileShape, float>;
+  using Multiply = cutlass::epilogue::fusion::Sm90Compute<
+      cutlass::multiplies,
+      float,
+      float,
+      cutlass::FloatRoundStyle::round_to_nearest>;
+  using Cast = cutlass::epilogue::fusion::Sm90Compute<
+      identity,
+      ElementOut,
+      float,
+      cutlass::FloatRoundStyle::round_to_nearest>;
+  using EpilogueEVT = cutlass::epilogue::fusion::Sm90EVT<
+      Cast,
+      cutlass::epilogue::fusion::Sm90EVT<
+          Multiply,
+          BScale,
+          cutlass::epilogue::fusion::Sm90EVT<Multiply, AScale, Accum>>>;
+
+  using CollectiveEpilogue =
+      typename cutlass::epilogue::collective::CollectiveBuilder<
+          cutlass::arch::Sm90,
+          cutlass::arch::OpClassSparseTensorOp,
+          TileShape,
+          cute::Shape<cute::_2, cute::_1, cute::_1>,
+          cutlass::epilogue::collective::EpilogueTileAuto,
+          ElementAccumulator,
+          float,
+          ElementOut,
+          cutlass::layout::RowMajor,
+          1,
+          ElementOut,
+          cutlass::layout::RowMajor,
+          1,
+          cutlass::epilogue::TmaWarpSpecializedCooperative,
+          EpilogueEVT>::CollectiveOp;
+
+  using CollectiveMainloop =
+      typename cutlass::gemm::collective::CollectiveBuilder<
+          cutlass::arch::Sm90,
+          cutlass::arch::OpClassSparseTensorOp,
+          ElementA,
+          cutlass::layout::RowMajor,
+          32,
```

Comment:
nit: would help with readability to define give these constant args variable names IMO